sklearn/impute.py

```diff
@@ -38,12 +39,20 @@
 ]
 
 
+def _custom_isnan(x):
+    # np.nan is never equal to np.nan. Return true only if x is np.nan
+    return x != x
+
+
 def _get_mask(X, value_to_mask):
     """Compute the boolean mask X == missing_values."""
     if value_to_mask is np.nan:
-        # nan values are never equal to themselves. We use this trick because
-        # np.isnan does not work on object dtypes.
-        return X != X
+        if X.dtype.kind in ("i", "u", "f"):
+            return np.isnan(X)
+        else:
+            # np.isnan does not work on object dtypes.
+            return custom_isnan(X)
```

Comment:
This if else clause could be part of `sklearn.utils.fixes.custom_isnan` so that it's fast for numerical data, even on old versions of numpy.